listenbrainz_spark/recommendations/candidate_sets.py

```diff
@@ -0,0 +1,363 @@
+
+        Args:
+            top_artists_with_collab (dataframe): Dataframe of top artists listened to by the user
+                whose similar artists count is not zero.
+            user_name (str): User name of the user.
+
+        Returns:
+            artists (dict): Dictionary can be depicted as:
+                {
+                    'top_artists 1' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                    'top_artists 2' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                    .
+                    .
+                    .
+                    'top_artists y' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                }
+    """
+    artists = defaultdict(dict)
+    for row in top_artist_with_collab.collect():
+        similar_artists_df = get_similar_artist_candidate_html(row.artist_name_0)
+        artists[row.artist_name_0] = [row.artist_name_1 for row in similar_artists_df.collect()]
+    return artists
+
+def save_candidate_html(user_data):
+    """ Save user data to an HTML file.
+
+        Args:
+            user_data (dict): Dcitionary can be depicted as:
+                {
+                    'user_name 1': {
+                        'artists': {
+                            'top_artists 1' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                        ...
+                        'top_artists y' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                        },
+                        'time' : 'xxx'
+                    },
+                }
+    """
+    date = datetime.utcnow().strftime('%Y-%m-%d')
+    candidate_html = 'Canidate-{}-{}.html'.format(uuid.uuid4(), date)
+    context = {
+        'user_data' : user_data
+    }
+    utils.save_html(candidate_html, context, 'candidate.html')
+
+def main():
+    ti = time()
+    try:
+        listenbrainz_spark.init_spark_session('Candidate_set')
+    except AttributeError as err:
+        logging.error('Cannot initialize Spark Session: {} \n {}. Aborting...'.format(type(err).__name__,
+            str(err)), exc_info=True)
+        sys.exit(-1)
+    except Exception as err:
+        logging.error('An error occurred: {} \n {}. Aborting...'.format(type(err).__name__,str(err)), exc_info=True)
+        sys.exit(-1)
+
+    listens_df = None
+    for y in range(config.STARTING_YEAR, config.ENDING_YEAR + 1):
+        for m in range(config.STARTING_MONTH, config.ENDING_MONTH + 1):
+            try:
+                month = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'
+                    .format(config.HDFS_CLUSTER_URI, y, m))
+                listens_df = listens_df.union(month) if listens_df else month
+            except AnalysisException as err:
+                logging.error('Cannot read parquet files from HDFS: {} \n {}'.format(type(err).__name__,str(err)))
```

Comment:
I think the error message should state which files could not be read.

Reply:
I have checked the error. 'str(err)' displays the path of files which could not be read, therefore I did not include them.